modPrecompiled/src/org/aion/precompiled/ContractExecutor.java

```diff
@@ -0,0 +1,298 @@
+        } else {
+            exeResult.setCodeAndNrgLeft(ResultCode.INTERNAL_ERROR, 0);
+        }
+    }
+
+    /**
+     * Finalize state changes and returns a summary.
+     *
+     * @return the execution summary.
+     */
+    private AionTxExecSummary finish() {
+
+        AionTxExecSummary.Builder builder = AionTxExecSummary.builderFor(getReceipt())
+            .logs(new ArrayList<>())
+            .deletedAccounts(new ArrayList<>())
+            .internalTransactions(new ArrayList<>())
+            .result(exeResult.getOutput());
+
+        switch (exeResult.getCode()) {
+            case SUCCESS:
+                repoTrack.flush();
+                break;
+            case INVALID_NONCE:
+            case INVALID_NRG_LIMIT:
+            case INSUFFICIENT_BALANCE:
+                builder.markAsRejected();
+                break;
+            case FAILURE:
+            case OUT_OF_NRG:
+            case REVERT:
+            case INTERNAL_ERROR:
+                builder.markAsFailed();
+                break;
+            default:
+                throw new RuntimeException("invalid code path, should not ever default");
+        }
+
+        AionTxExecSummary summary = builder.build();
+
+        if (!isLocalCall && !summary.isRejected()) {
+            IRepositoryCache track = repo.startTracking();
+            // refund nrg left
+            if (exeResult.getCode() == ResultCode.SUCCESS || exeResult.getCode() == ResultCode.REVERT) {
+                track.addBalance(tx.getFrom(), summary.getRefund());
+            }
+
+            tx.setNrgConsume(tx.nrgLimit() - exeResult.getNrgLeft());
+
+            // Transfer fees to miner
+            track.addBalance(block.getCoinbase(), summary.getFee());
+
+            track.flush();
+        }
+
+        if (contractLogger.isDebugEnabled()) {
+            contractLogger.debug("Transaction receipt: {}", summary.getReceipt());
+            contractLogger.debug("Transaction logs: {}", summary.getLogs());
+        }
+
+        return summary;
+    }
+
+    /**
+     * Returns the transaction receipt.
+     *
+     * @return the transaction receipt.
+     */
+    protected AionTxReceipt getReceipt() {
+        AionTxReceipt receipt = new AionTxReceipt();
+        receipt.setTransaction(tx);
+        receipt.setLogs(new ArrayList<>());
+        receipt.setNrgUsed(tx.nrgLimit() - exeResult.getNrgLeft());
+        receipt.setExecutionResult(exeResult.getOutput());
+        receipt.setError(exeResult.getCode() == ResultCode.SUCCESS ? "" : exeResult.getCode().name());
+        return receipt;
+    }
+
+    /**
+     * Tells the ContractExecutor to bypass incrementing the account's nonce when execute is called.
+     */
+    public void setBypassNonce() {
+        this.askNonce = false;
+    }
+
+    /**
+     * Tells the ContractExecutor not to bypass incrementing the account's nonce when execute is
+     * called.
+     */
+    public void setNoBypassNonce() {
+        this.askNonce = true;
```

Comment:
why not make it a single method with a parameter like in the `TransactionExecutor`?